google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1alpha2/JsonStreamWriter.java

```diff
@@ -0,0 +1,359 @@
+  // //   data.setRows(rowsBuilder.build());
+  // //
+  // //   ApiFuture<AppendRowsResponse> appendResponseFuture =
+  // // this.streamWriter.append(AppendRowsRequest.newBuilder().setProtoRows(data.build()).build());
+  // //   // AppendRowsResponse appendResponse = appendResponseFuture.get();
+  // //   // if (appendResponse.hasUpdatedSchema()) {
+  // //   //
+  // //   //   // this.descriptor =
+  // // BQTableSchemaToProtoDescriptor.convertBQTableSchemaToProtoDescriptor(BQTableSchema);
+  // //   // }
+  //   // return appendResponseFuture;
+  // }
+
+  /**
+   * Updates the descriptor and BQTable schema. This function is used to support schema updates, and
+   * is called whenever AppendRowsResponse includes an updated schema. The function is synchronized
+   * since it is called through a callback (which is in another thread). If the main thread is
+   * calling append while the callback thread calls refreshAppend(), this might cause some issues.
+   *
+   * @param BQTableSchema The updated table schema.
+   */
+  private synchronized void updateDescriptor(Table.TableSchema updatedSchema)
+      throws Descriptors.DescriptorValidationException, IOException, InterruptedException {
+    if (!this.BQTableSchema.equals(updatedSchema)) {
+      this.descriptor =
+          BQTableSchemaToProtoDescriptor.convertBQTableSchemaToProtoDescriptor(updatedSchema);
+      this.BQTableSchema = updatedSchema;
+      streamWriter.refreshAppend();
+    }
+  }
+
+  /** Placeholder. Will update to include JsonToProtoMessage conversion. */
+  public synchronized ApiFuture<AppendRowsResponse> append(AppendRowsRequest message) {
+    ApiFuture<AppendRowsResponse> appendResponseFuture = this.streamWriter.append(message);
+    ApiFutures.<AppendRowsResponse>addCallback(
+        appendResponseFuture,
+        new ApiFutureCallback<AppendRowsResponse>() {
+          @Override
+          public void onSuccess(AppendRowsResponse response) {
+            if (response.hasUpdatedSchema()) {
+              try {
+                updateDescriptor(response.getUpdatedSchema());
+              } catch (Exception e) {
+                // Deal with exceptions
+              }
+            }
+          }
+
+          @Override
+          public void onFailure(Throwable t) {}
+        });
+    return appendResponseFuture;
+  }
+
+  /**
+   * Gets streamName
+   *
+   * @return String
+   */
+  public String getStreamName() {
+    return this.streamName;
+  }
+
+  /**
+   * Gets current descriptor
+   *
+   * @return Descriptor
+   */
+  public Descriptor getDescriptor() {
+    return this.descriptor;
+  }
+
+  /**
+   * Gets current BQTableSchema
+   *
+   * @return Table.TableSchema
+   */
+  public Table.TableSchema getBQTableSchema() {
+    return this.BQTableSchema;
+  }
+
+  /** Sets all StreamWriter settings. */
+  private void setStreamWriterSettings(
+      StreamWriter.Builder builder,
```

Comment:
Mark these as nullable?

Reply:
Done.

Reply:
Better to make them optional instead of nullable.

Reply:
Done.